### PR TITLE
Fix: Sort portal types alphabetically in search filter

### LINKED ISSUE #3860: List of portal types in content-type filter unsupported
Plone 6.0.7

The list of available portal types in the filter should/must be sorted alphabetically.

Talking of sorting: the standard `sort()` or `sorted()` method should not be applied because they do not deal corr

ectly with unicode strings in the context of the configured locale using `pyuca`:

https://github.com/jtauber/pyuca

d0f18fc9-40a1-465b-a7db-d5b68cdfb8a8

### PR DESCRIPTION
**Pull request by rohnsha0**
fixes #3860

### CODE REVIEW

**Comment by rohnsha0** [comment]
@jenkins-plone-org please run jobs

File: Products/CMFPlone/browser/search.py
Comment (MrTango) [inline Products/CMFPlone/browser/search.py:192]: > Talking of sorting: the standard sort() or sorted() method should not be applied because they do not deal correctly with unicode strings in the context of the configured locale using pyuca:
> https://github.com/jtauber/pyuca

please take the comment in the referenced issue into account ;)

**Review by MrTango — CHANGES_REQUESTED** [review]

File: Products/CMFPlone/browser/search.py
Comment (petschki) [reply Products/CMFPlone/browser/search.py:192]: problem here is, that this only sorts the internal stored `portal_type` (eg. `Document`) but gets later translated in the template to `Seite` (german) ... this needs to be refactored to a translated and sorted vocabulary with internal type as key and translated `type_info.Title()` as label

**Review by petschki — CHANGES_REQUESTED** [review]

File: Products/CMFPlone/browser/search.py
Comment (yurj) [reply Products/CMFPlone/browser/search.py:192]: [pyuca](https://github.com/jtauber/pyuca) is 7 years old, use [pyICU](https://gitlab.pyicu.org/main/pyicu) instead (if possible).
```python
>>> from icu import RuleBasedCollator
>>> collator = RuleBasedCollator(
...     """
...     &A<ą<<<Ą
...     &C<ć<<<Ć
...     &E<ę<<<Ę
...     &L<ł<<<Ł
...     &N<ń<<<Ń
...     &O<ó<<<Ó
...     &S<ś<<<Ś
...     &Z<ź<<<Ź<ż<<<Ż
...     """
... )
>>> sorted(polish_names, key=collator.getSortKey)
['Ludmiła', 'Łukasz', 'Zbigniew', 'Żaneta']
```

You could also use locale() sorting capability as stated here: https://realpython.com/python-sort-unicode-strings/#leverage-pythons-locale-module-to-sort-unicode-strings
```python
>>> locale.setlocale(locale.LC_COLLATE, "pl_PL.UTF-8")
'pl_PL.UTF-8'
>>> sorted(polish_names, key=locale.strxfrm)
['Ludmiła', 'Łukasz', 'Zbigniew', 'Żaneta']
```
but beware of potential problems setting the locale, and also you need all the language locales set in your system.

Note: do this need a test too?

Comment (petschki) [reply Products/CMFPlone/browser/search.py:192]: I'd say: use the existing vocabulary for the searchfilter here and take care of sorting in `plone.app.vocabulary` -> new PR ...

Comment (zopyx) [reply Products/CMFPlone/browser/search.py:192]: > The problem of pyICU is the scarse availability as compiled wheels in pypi, so we would depend on this. Windows has builds: https://github.com/cgohlke/pyicu-build

Attention with pyICU because it requires the related library to be installed on your OS. It's is no installed by default and would require an update of the Plone installation instructions for the related OS or distribution...also not sure about compatibility/availability of ICU on Windows.

Comment (rohnsha0) [reply Products/CMFPlone/browser/search.py:192]: I'm stuck. I don't know how to make those changes. @petschki, @zopyx, and @yurj, can you please point me in the right direction?

Comment (zopyx) [reply Products/CMFPlone/browser/search.py:192]: Point is that need one of the following solutions which basically can not directly integrated:
- `locale.setlocatel()` - an anti-pattern for switching the locale while an app is running
- `pyuca` - likely too old and unmaintained (but working for me under Python 3.12, but a new dependency)
- `pyICU`- primary choice but a new huge dependency that is likely unwanted as a dependency of Plone

Sorting the portal_types based on their translated title should be good enough in most case. I doubt that you need collation (as proposed using `pyuca`, `pyICU`) for the few portal type strings.

**Review by petschki — CHANGES_REQUESTED** [review]
I have doubts about this change:

- its not the right place here. We should use the already translated and sorted vocabulary from `plone.app.vocabularies.types.ReallyUserFriendlyTypesVocabulary`
- using `icu` for sorting simple portal type strings is a bit overengineered IMHO

**Comment by rohnsha0** [comment]
@petschki I am unsure tho about how to implement it here! Any help would be greatly appreciated!

**Comment by petschki** [comment]
I found out, that the list in the `@@search-controlpanel` already uses the mentioned vocabulary. Take a look here https://github.com/plone/Products.CMFPlone/blob/master/Products/CMFPlone/controlpanel/browser/search.py how the vocabulary is fetched.

**Comment by rohnsha0** [comment]
i testedd the latest changes and it worked... please review changes @petschki

**Comment by rohnsha0** [comment]
@jenkins-plone-org please run jobs

**Comment by davisagli** [comment]
Done in https://github.com/plone/plone.app.vocabularies/pull/93 instead

**Comment by davisagli** [comment]
Sorry, I misunderstood -- this PR is still needed in order to use the change from https://github.com/plone/plone.app.vocabularies/pull/93

File: Products/CMFPlone/browser/templates/search.pt
Comment (davisagli) [inline Products/CMFPlone/browser/templates/search.pt:90]: What is this for?

**Comment by rohnsha0** [comment]
> Sorry, I misunderstood -- this PR is still needed in order to use the change from [plone/plone.app.vocabularies#93](https://github.com/plone/plone.app.vocabularies/pull/93)

exactly...

File: Products/CMFPlone/browser/templates/search.pt
Comment (rohnsha0) [reply Products/CMFPlone/browser/templates/search.pt:90]: adds a minor padding btn `Filter the results` & `Sort by` options...
not neccessary for the functionality at all.. just was looking odd so added it...

image

**Comment by rohnsha0** [comment]
@jenkins-plone-org please run jobs

**Review by davisagli — APPROVED** [review]

**Comment by rohnsha0** [comment]
@MrTango @petschki please verify the requested changes....

File: Products/CMFPlone/browser/search.py
Comment (wesleybl) [inline Products/CMFPlone/browser/search.py:189]: I find the `_portal_types_vocabulary` method unnecessary.

File: Products/CMFPlone/browser/search.py
Comment (wesleybl) [inline Products/CMFPlone/browser/search.py:190]: I think it's better to leave this test out. If someone removes the utility, the CI tests better fail and the person who removes it better analyzes the consequences of the removal.

File: Products/CMFPlone/browser/search.py
Comment (wesleybl) [inline Products/CMFPlone/browser/search.py:189]: Why did you remove this comment?

File: Products/CMFPlone/browser/search.py
Comment (wesleybl) [reply Products/CMFPlone/browser/search.py:190]: You can test both PRs together at:

https://jenkins.plone.org/job/pull-request-6.1-3.10/build?delay=0sec

**Comment by stevepiercy** [comment]
The issue this PR is supposed to fix was fixed elsewhere. See https://github.com/plone/Products.CMFPlone/issues/3860#issuecomment-2246819003

**Comment by rohnsha0** [comment]
> The issue this PR is supposed to fix was fixed elsewhere. See [#3860 (comment)](https://github.com/plone/Products.CMFPlone/issues/3860#issuecomment-2246819003)

@stevepiercy please see https://github.com/plone/Products.CMFPlone/pull/3985#issuecomment-2246824261

**Comment by stevepiercy** [comment]
Sorry for my mistake.

File: Products/CMFPlone/browser/search.py
Comment (jensens) [reply Products/CMFPlone/browser/search.py:189]: yes, just add `queryUtility` here - do not blow up blown up code more than needed.

File: Products/CMFPlone/browser/search.py
Comment (jensens) [reply Products/CMFPlone/browser/search.py:189]: +1 please keep it.

Comment (rohnsha0) [reply Products/CMFPlone/browser/search.py:189]: ah, that might be a mistake... apologies

File: Products/CMFPlone/browser/search.py
Comment (rohnsha0) [reply Products/CMFPlone/browser/search.py:190]: okay.. removed the test

File: Products/CMFPlone/browser/search.py
Comment (rohnsha0) [reply Products/CMFPlone/browser/search.py:189]: done

File: Products/CMFPlone/browser/search.py
Comment (wesleybl) [inline Products/CMFPlone/browser/search.py:18]: Please put all imports in alphabetical order.

**Review by wesleybl — CHANGES_REQUESTED** [review]

File: Products/CMFPlone/browser/search.py
Comment (wesleybl) [inline Products/CMFPlone/browser/search.py:195]: Why is `set` necessary here?

File: Products/CMFPlone/browser/search.py
Comment (rohnsha0) [reply Products/CMFPlone/browser/search.py:17]: done

File: Products/CMFPlone/browser/search.py
Comment (rohnsha0) [reply Products/CMFPlone/browser/search.py]: reverted

**Comment by rohnsha0** [comment]
@jenkins-plone-org please run jobs

File: news/3860.bugfix
Comment (wesleybl) [inline news/3860.bugfix:1]: Period at the end of the phase.

**Review by wesleybl — CHANGES_REQUESTED** [review]

File: news/3860.bugfix
Comment (stevepiercy) [inline news/3860.bugfix:1]: For languages that don't have an alphabet. Also we should refer to the implementation in the other package. And use `@` for credit.

```suggestion
Sort portal types in search filter according to `ReallyUserFriendlyTypes` using `unidecode` in `plone.app.vocabularies`. @rohnsha0
```

File: news/3860.bugfix
Comment (stevepiercy) [inline news/3860.bugfix:2]:
```suggestion

```

**Review by stevepiercy — CHANGES_REQUESTED** [review]
Just a change on the news item, and it's good to merge. Thank you!

File: news/3860.bugfix
Comment (rohnsha0) [reply news/3860.bugfix]: ah... gotcha... thanks btw

**Review by wesleybl — APPROVED** [review]
LGTM

**Comment by rohnsha0** [comment]
@jenkins-plone-org please run jobs

**Comment by rohnsha0** [comment]
> @rohnsha0 It would be good to make these changes in the 6.0.x branch as well.

pushing these changes directly to that branch too!

**Comment by rohnsha0** [comment]
> > @rohnsha0 It would be good to make these changes in the 6.0.x branch as well.
> 
> pushing these changes directly to that branch too!

got below errors while pushing the changes
```
remote: error: GH006: Protected branch update failed for refs/heads/6.0.x.
remote: error: Changes must be made through a pull request. 2 of 2 required status checks are expected.
```

**Comment by wesleybl** [comment]
@rohnsha0 You need to create a new branch based on the 6.0.x branch and create a new PR.

**Comment by stevepiercy** [comment]
@rohnsha0 `6.0.x` is a release branch. We should never push directly to a release branch. A separate PR is required, what we call a "back port". You can do that after this PR is approved and merged.

**Review by stevepiercy — APPROVED** [review]

**Comment by stevepiercy** [comment]
@jenkins-plone-org please run jobs

**Comment by stevepiercy** [comment]
As soon as all checks are green, let's merge. Thank you!

**Comment by wesleybl** [comment]
@gforcada  It seems that Jenkins `Node3` has a problem testing with Python 3.12. To see:

https://jenkins.plone.org/job/pull-request-6.1-3.12/507/console

**Comment by stevepiercy** [comment]
Does anyone know why Jenkins is choking on set up? I've been seeing these types of errors frequently. @gforcada @mauritsvanrees 

https://jenkins.plone.org/job/pull-request-6.1-3.12/507/consoleFull

```
Setup failed:
SessionNotCreatedException: Message: session not created
from unknown error: cannot create temp dir for user data dir
```

**Comment by wesleybl** [comment]
> Does anyone know why Jenkins is choking on set up? I've been seeing these types of errors frequently.

Whatever it is, it appears to only occur on Node3.

**Comment by rohnsha0** [comment]
> @rohnsha0 `6.0.x` is a release branch. We should never push directly to a release branch. A separate PR is required, what we call a "back port". You can do that after this PR is approved and merged.

Ah... Thanks for the insight... Was unaware 😬

**Comment by wesleybl** [comment]
> Setup failed:
SessionNotCreatedException: Message: session not created
from unknown error: cannot create temp dir for user data dir

This error is usually caused when the chromedriver version is not compatible with the chrome version. See:

https://developer.chrome.com/docs/chromedriver/downloads

Anyway, I think this PR can be merged before fixing this. Jenkins is green.

**Comment by stevepiercy** [comment]
Let's get a final review from either @petschki or @MrTango for the recent changes since their initial reviews. If they don't respond in a week or two, or we get another approval from a maintainer before then, then we can merge.

**Comment by rohnsha0** [comment]
@jenkins-plone-org please run jobs

File: Products/CMFPlone/browser/search.py
Comment (jensens) [inline Products/CMFPlone/browser/search.py:198]: I think the naming and all is that confusing, so a good amount of explanation is good for our future selfs.

```suggestion

        # We want to have the configured types to be exposed in the search sorted for humans (by translated Title).     
        # Those are stored in the Plone registry. They are called here UserFriendlyTypes.        
        # 
        # Confusingly, on the other hand we have the ReallyUserFriendlyTypes vocabulary from 
        # plone.app.vocabularies, which is already sorted accordingly and contain all possible types, 
        # except "Temp Folder" and "Plone Site". 

        # fetch the sorted ReallyUserFriendlyTypes vocabulary
          vocab_factory = queryUtility(
            IVocabularyFactory, "plone.app.vocabularies.ReallyUserFriendlyTypes"
        )
        vocab = vocab_factory(self.context)

        # get the configured values from the registry and pass the input types to be reduced to possible values
        user_friendly_types = plone_utils.getUserFriendlyTypes(types)

        # Filter the sorted ReallyUserFriendlyTypes down to the configured values from the registry,
        #  but keep the order.
        sorted_types = [term.value for term in vocab if term.value in user_friendly_types]

        return sorted_types
```

**Comment by rohnsha0** [comment]
@jenkins-plone-org please run jobs

**Comment by rohnsha0** [comment]
I have literally zero idea why is Jenkins failing! @jensens

**Review by jensens — APPROVED** [review]

**Comment by rohnsha0** [comment]
@jenkins-plone-org please run jobs

**Comment by stevepiercy** [comment]
@rohnsha0 I visited the Jenkins site. I've observed that Node 1 has been flaky for the last couple of weeks, where all of its tests fail.

I compared the full console output of a [successful run on Node 3](https://jenkins.plone.org/job/pull-request-6.1-3.10/277/consoleFull) versus a [failed run on Node 1](https://jenkins.plone.org/job/pull-request-6.1-3.10/280/consoleFull), and found that it appears the server against which `zope.testrunner` runs its tests never starts.

### Successful run

```
+ bin/test --all --xml . -t '!ONLYROBOT'
Running zope.testrunner.layer.UnitTests tests:
  Set up zope.testrunner.layer.UnitTests in 0.000 seconds.
127.0.0.1 - - [07/Aug/2024 23:51:34] "PURGE /foo HTTP/1.1" 200 -
```

### Failed run

```
+ bin/test --all --xml . -t '!ONLYROBOT'
Running zope.testrunner.layer.UnitTests tests:
  Set up zope.testrunner.layer.UnitTests in 0.000 seconds.


Failure in test /home/jenkins/.buildout/eggs/cp310/plone.app.testing-7.1.0-py3.10.egg/plone/app/testing/layers.rst
Failed doctest test for layers.rst
  File "/home/jenkins/.buildout/eggs/cp310/plone.app.testing-7.1.0-py3.10.egg/plone/app/testing/layers.rst", line 0

<snip>

Tearing down left over layers:
  Tear down plone.testing.zope.Startup in 0.002 seconds.
  Tear down plone.testing.zca.LayerCleanup in 0.000 seconds.
Total: 0 tests, 0 failures, 7 errors and 0 skipped in 1.165 seconds.
+ return_code=1
+ '[' 1 = all_right ']'
+ python templates/pr-update-status-py3.py
+ exit 1
Build step 'Execute shell' marked build as failure
```

I don't know how to interpret these results. I hope someone familiar with Jenkins can figure it out and enlighten us. Is this an @plone/ai-team issue? I don't know where to report this.

**Comment by rohnsha0** [comment]
> @rohnsha0 I visited the Jenkins site. I've observed that Node 1 has been flaky for the last couple of weeks, where all of its tests fail. 

and its a real pain, which needs to be fixed. cant we just redirect its workload to other available node?

**Comment by stevepiercy** [comment]
Redirecting merely avoids the issue. It fails more than succeeds, and it appears to be only for robot tests. It appears to have first started to fail with build #207.

https://jenkins.plone.org/job/pull-request-6.1-3.10/277/robot/

image

I assume it's up to the Admin and Infrastructure Team to resolve.

Meanwhile, we can ignore the false failure, and merge this PR. It [passed on Node 1](https://jenkins.plone.org/job/pull-request-6.1-3.10/270/console), the only changes after that were comments, and @jensens reviewed and approved your work. I think you have permission as a member of the Developer Team to check the box for `Merge without waiting for requirements to be met (bypass branch protections)`, then merge.

**Review by mauritsvanrees — APPROVED** [review]
I will merge.  Thanks!

**Comment by mauritsvanrees** [comment]
> > @rohnsha0 I visited the Jenkins site. I've observed that Node 1 has been flaky for the last couple of weeks, where all of its tests fail.
> 
> and its a real pain, which needs to be fixed. cant we just redirect its workload to other available node [Node3]?

Actually, from what I remember the last weeks, it has been Node 3 that gives most problems. :-/

**Comment by rohnsha0** [comment]
> Redirecting merely avoids the issue. It fails more than succeeds, and it appears to be only for robot tests. It appears to have first started to fail with build https://github.com/plone/Products.CMFPlone/issues/207.

actually yes... in this pr itslef, it failed multiple times

**Comment by rohnsha0** [comment]
> > > @rohnsha0 I visited the Jenkins site. I've observed that Node 1 has been flaky for the last couple of weeks, where all of its tests fail.
> > 
> > 
> > and its a real pain, which needs to be fixed. cant we just redirect its workload to other available node [Node3]?
> 
> Actually, from what I remember the last weeks, it has been Node 3 that gives most problems. :-/

correct...

**Comment by stevepiercy** [comment]
Node 3 has been failing, too, but not as much as Node 1.

I looked again at the full log. These snippets may be interesting.

```
AssertionError: Setup failed:
SessionNotCreatedException: Message: session not created
from unknown error: cannot create temp dir for user data dir
```

```
Also teardown failed:
No browser is open.
```

We saw similar behavior in Volto for the last few weeks with the browser not starting and Cypress tests would fail. We recently upgraded it to the latest version. Maybe bumping beyond `robotframework==6.0.2` will help? I didn't see anything relevant in its change log, though.

**Comment by mauritsvanrees** [comment]
I see an actual error in the code snuck in with the final changes here.  If you look at the [full console log](https://jenkins.plone.org/job/plone-6.1-python-3.10/lastCompletedBuild/consoleFull) of a job after this merge, you see a traceback:

```
        IndentationError: unindent does not match any outer indentation level
                                         ^

            vocab_factory = queryUtility(

          File "/home/jenkins/workspace/plone-6.1-python-3.10/src/Products.CMFPlone/Products/CMFPlone/browser/search.py", line 196
```

In fact, Plone itself does not even start up now.
I have [pushed a fix](https://github.com/plone/Products.CMFPlone/commit/c071c8c893cebf34d947eb8b9d7546cb60cdd6d5) directly to master.

That's the problem when Jenkins is already flaky to begin with, then errors like this may slip through the cracks.

**Comment by rohnsha0** [comment]
> I saw that, and thought that error was from sloppy reStructuredText indentation, not Python, because that section starts with:
> 
> ```
> File "/home/jenkins/.buildout/eggs/cp310/plone.app.testing-7.1.0-py3.10.egg/plone/app/testing/layers.rst", line 47, in layers.rst
> ```
> 
> I now see that I read the stacktrace incorrectly.
> 
> Would [running tests locally](https://6.docs.plone.org/contributing/core/index.html#test-locally) (now that it is documented well for Plone 6) before pushing the commit have caught it? Asking for a friend...

actually i tested the code locally before pushing and that worked. but ig yesterday we did the documentation which I directly committed from web... so maybe that went wrong... Sorry, Apologies @mauritsvanrees @stevepiercy

**Comment by mauritsvanrees** [comment]
Yes, that would have caught it.
Also, just starting the instance would have caught it, which apparently neither the committer, nor the last reviewer (me) did...
Not blaming anyone here, mistakes happen.

**Comment by rohnsha0** [comment]
> Yes, that would have caught it. Also, just starting the instance would have caught it, which apparently neither the committer, nor the last reviewer (me) did... Not blaming anyone here, mistakes happen.

yep! nevermind... would test every changes before pushing tho.. hahaaa 🤐

**Pull request merged**
